src/core/src/engine/fork_resolver.rs

```diff
@@ -14,229 +14,302 @@
-            state_store.delete(next.clone());
-            next = BlockId::from(state_.unwrap().estimate_info.previous_block_id.as_bytes().to_vec());
+
+        else if chain_block.block_num < fork_block.block_num {
+            // Keep getting blocks from fork until same height
+            // as the chain is reached.
+            while chain_block.block_num != fork_block.block_num {
+                // Keep adding wait times
+                // Get cc from certificate in fork_block
+                self.fork_cc += self.get_wait_time_from(&fork_block);
+                self.fork_len += 1;
+                match service.get_block(&fork_block.previous_id) {
+                    Ok( ancestor ) => {
+                        fork_block = ancestor;
+                    }
+                    Err(err) => {
+                        error!("Error getting block from validator");
+                        break;
+                    }
+                }
+            }
         }
-    }
-}
 
-fn get_cert_from(block: &Block) -> WaitCertificate {
-    let payload = block.payload.clone();
-    debug!("Extracted payload from block: {:?}", payload.clone());
-    let (wait_certificate, _) = poet2_util::payload_to_wc_and_sig(&payload);
-    debug!("Serialized wait_cert : {:?}", &wait_certificate);
-    serde_json::from_str(&wait_certificate).unwrap()
+        // Loop over fork and chain to find a common ancestor
+        // Descend to the ancestor/previous block in each
+        // iteration for both the chain & the fork.
+        // If genesis is reached in the process( which is almost
+        // not possible), stop iteration and a completely new
+        // chain is competing for fork resolution.
+        loop {
+
+            let prev_chain_block_id = chain_block.previous_id.clone();
+            let prev_fork_block_id = fork_block.previous_id;
+            if prev_fork_block_id == prev_chain_block_id {
+                // Found common ancestor
+                info!("Found a common ancestor at block_id {}",
+                    poet2_util::to_hex_string(&prev_fork_block_id));
+                ancestor_found = true;
+                break;
+            }
+            let blocks_map = service
+                .get_blocks(vec![prev_chain_block_id.clone(),
+                             prev_fork_block_id.clone()]);
+            match blocks_map {
+                Err(err) => {
+                    error!("Could not get blocks");
+                    ancestor_found = false;
+                    break;
+                }
+                Ok(mut block_map) => {
+                    // Remove from the returned hashmap to get block
+                    chain_block = block_map.remove(&prev_chain_block_id)
+                        .expect("Could not extract block from map.");
+                    self.chain_len += 1;
+                    // Get cc from certificate in chain_block
+                    self.chain_cc += self.get_wait_time_from(&chain_block);
```

Comment:
This value should be obtained from consensus state store. If not present extract from wait certificate and add it to state store